File: algorithms/recursion/power.js
var isEven = function(n) {
    return n % 2 === 0;
};

var isOdd = function(n) {
    return !isEven(n);
};

var power = function(x, n) {
    println("Computing " + x + " raised to power " + n + ".");
    // base case
    if (n === 0) {
        return 1;
    }
    if (n > 0) {
        if (isEven(n)) {
            var result = power(x, n/2);
            return result * result;
        } else if (isOdd(n)) {
            return x * power(x, n - 1);
        }
    }
    if (n < 0) {
        return 1 / power(x, -n);
    }
};

var displayPower = function(x, n) {
    println(x + " to the " + n + " is " + power(x, n));
};

displayPower(3, 0);
Program.assertEqual(power(3, 0), 1);
displayPower(3, 1);
Program.assertEqual(power(3, 1), 3);
displayPower(3, 2);
Program.assertEqual(power(3, 2), 9);
displayPower(3, -1);
Program.assertEqual(power(3, -1), 1/3);
displayPower(4, 2);
Program.assertEqual(power(4, 2), pow(4, 2));
displayPower(2, 4);
Program.assertEqual(power(2, 4), pow(2, 4));

File: algorithms/recursion/power.py
def power(x, n):
    print("Computing ", x, " raised to power ", n, ".")
    if n == 0: # 지수가 0일 경우
        return 1
    elif n > 0: # 지수가 양수일 경우
        if n % 2 == 0: # 지수가 짝수일 경우
            return power(x, n/2) * power(x, n/2)
        else: # 지수가 홀수일 경우
            return x * power(x, n-1)
    else: # 지수가 음수일 경우
        return 1 / power(x, -n)

if __name__ == "__main__":
    print(power(2, 3))
